### ISSUE TITLE
ToT / 14 does not build on OpenBSD/amd64

### ISSUE THREAD

**Issue by brad0** [llvm/llvm-project#53614]
At the moment  the ToT / 14 will not build on OpenBSD/amd64. I ran into this a few months ago and
then forgot about it.

It seems to be caused by this commit.. https://github.com/llvm/llvm-project/commit/a0a5964499816373c50d6d6a3a4b38c1b53f6714
https://reviews.llvm.org/D105466

I have e-mailed some of our developers to see if they can comment on the diff and why it might be failing as is.

```
ld: error: undefined symbol: LLVMRTDyldTLSSpace
>>> referenced by llvm-rtdyld.cpp
>>>               tools/llvm-rtdyld/CMakeFiles/llvm-rtdyld.dir/llvm-rtdyld.cpp.o:(TrivialMemoryManager::allocateTLSSection(unsigned long, unsigned int, unsigned int, llvm::StringRef))
c++: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by tstellar** [comment]
@MoritzS @lhames Should we revert this? https://github.com/llvm/llvm-project/commit/a0a5964499816373c50d6d6a3a4b38c1b53f6714

**Comment by lhames** [comment]
@brad0 Does adding `__linux__` to the condition fix the issue? If so I think we should just do that for now

@MoritzS Do you need this to work for BSD too? If so, would you be willing to investigate a longer term fix?

**Comment by MoritzS** [comment]
I used `#if defined(__x86_64__) && defined(__ELF__)` as a proxy to test for whether the target uses the ELF x86-64 TLS ABI. I'm not familiar with OpenBSD, does it use a different ABI? If adding `__linux__` fixes this, I would also be in favor to just do that.

**Comment by brad0** [comment]
I'll check to see if that builds when I get home in a number of hours. But just to add to this.. OpenBSD currently does not have native TLS support. We're using the emulated TLS support.

**Comment by brad0** [comment]
> I used `#if defined(__x86_64__) && defined(__ELF__)` as a proxy to test for whether the target uses the ELF x86-64 TLS ABI. I'm not familiar with OpenBSD, does it use a different ABI? If adding `__linux__` fixes this, I would also be in favor to just do that.

I changed `#if defined(__x86_64__) && defined(__ELF__)` to `#if defined(__x86_64__) && defined(__ELF__) && !defined(__OpenBSD__)` and it built Ok.

**Comment by MoritzS** [comment]
Ok, that's good to know. Thanks for testing, @brad0. To be safe I think we should use `defined(__linux__)` instead of `!define(__OpenBSD__)`. I only tested this extensively on linux and even the test case itself only runs on linux.

**Comment by brad0** [comment]
> Ok, that's good to know. Thanks for testing, @brad0. To be safe I think we should use `defined(__linux__)` instead of `!define(__OpenBSD__)`. I only tested this extensively on linux and even the test case itself only runs on linux.

Ah, it's disabling the code either way on OpenBSD. I had suggested !OpenBSD thinking other OS's like FreeBSD / NetBSD for example were  Ok for TLS and ABI wise.

**Comment by brad0** [comment]
@MoritzS ?

**Comment by brad0** [comment]
Posted a diff here for review.. https://reviews.llvm.org/D119989

**Comment by lhames** [comment]
@brad0 -- that diff looks like it's for an unrelated issue?

Since @MoritzS was happy with a `defined(__linux__)` guard I will implement that for now.

**Comment by tstellar** [comment]
Merged: 56ac6dbc7369
